### PR TITLE
Translate host suffixes in no_proxy to JVM syntax

### PR DESCRIPTION
**Pull request by cruhland**
This ended up being a bigger change than intended. I decided to write property tests for the `no_proxy` parsing to make sure all the edge cases got covered. But to make it easier to write the tests, I first separated the initialization code in `HttpProxySupport` into two parts:

1. parse the values of the system properties (without side effects)
2. set the values in the global system properties

### CODE REVIEW

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/HttpProxySupport.scala [hunk 1]
```diff
@@ -102,43 +102,50 @@ private[cosmos] object HttpProxySupport {
...
-        }
+  private[cosmos] def parseHostPortProperties(
+    hostProperty: String,
+    portProperty: String
+  )(proxyUri: Uri): List[(String, String)] = {
+    Uris.extractHostAndPort(proxyUri)
+      .toOption
+      .toList
+      .flatMap { case ConnectionDetails(host, port, _) =>
+        List((hostProperty, host), (portProperty, port.toString))
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/HttpProxySupport.scala]: I am okay making all of the "functional" method public if you need them for test while keeping all of the side-effecting helper methods `private[this]`.

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala [hunk 2]
```diff
@@ -25,18 +25,29 @@ final class HttpProxySupportSpec extends FreeSpec with BeforeAndAfter {
...
+
+        "respects http.proxyHost and http.proxyPort" in {
           System.setProperty(HttpHost, "host")
           System.setProperty(HttpPort, "port")
-          HttpProxySupport.initHttpProxyProperties(Some(proxyUri), HttpHost, HttpPort)
+
+          val httpProxyUri = "http://localhost:3128"
+          val proxyEnvVariables = ProxyEnvVariables(Some(httpProxyUri), None, None)
+          HttpProxySupport.initProxyConfig(proxyEnvVariables, _ => ())
+
```
Comment (jsancio) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: I see. Maybe you should ignore my `private[this]` vs `private[cosmos]` comment since we have this side-effecting change. In this case `private[cosmos]` is almost public. There are a bunch of unrelated classes in `com.mesosphere.cosmos`.

Actually, if I understand this test correctly, this test looks dangerous. Aren't we overriding the system properties?

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala [hunk 3]
```diff
@@ -141,6 +141,12 @@ final class HttpProxySupportSpec extends FreeSpec with BeforeAndAfter {
             }
           }
 
+          "should add wildcard in front of '.'-delimited domain suffixes" in {
+            assertResult("*.mesos|*.example.com") {
+              HttpProxySupport.translateNoProxy(".mesos,.example.com")
+            }
+          }
+
```
Comment (jsancio) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: Should we add a test for IPs? E.g `...translateNoProxy(".0.0.100") shouldBe "*.0.0.100"`. Is that even allowed? We can leave it as a known "bug".

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala [hunk 4]
```diff
@@ -297,4 +291,47 @@ object HttpProxySupportSpec {
...
+      .map { host =>
+        val prefix = host.wildcard match {
+          case NoWildcard => ""
+          case _ => "*."
+        }
+
+        prefix + host.elements.mkString(".")
+      }
+      .mkString("|")
+  }
```
Comment (jsancio) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: Very cool. Having said that explicit tests are easier to read. Maybe we want both. Static tests like we had before and generated tests.

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala (hunk 2 above)
Comment (cruhland) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: Yeah, these tests were written this way before my change, and it's a unit test so I'm not as concerned about modifying system properties. What do you think should be done differently?

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala (hunk 4 above)
Comment (cruhland) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: Good point, I'll bring back the static tests.

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala (hunk 2 above)
Comment (cruhland) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: I mean, I could try to isolate the side effects to a single method, and update most of the unit tests to use pure helper methods instead. Then we'd still need a side-effecting test, but it would be very restricted.

Comment (jsancio) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: Yeah. We don't need to change the test. I was just pointed it out. I didn't notice it when the test was first introduced.

File: cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala (hunk 3 above)
Comment (cruhland) [inline cosmos-test-common/src/test/scala/com/mesosphere/cosmos/HttpProxySupportSpec.scala]: Oops, missed this comment. I agreed that we don't need to worry about this case until it happens in the wild. Even then, I'm not sure we should support it -- IPs should always be explicit.

**Comment by larioj** [comment]
@cruhland It seems to me that the other PRs are passing with some frequency. How many time have you rerun this?

**Comment by cruhland** [comment]
@larioj probably about ten times by now...

**Comment by cruhland** [comment]
It's not consistently the same test that fails, but it's definitely the same error. Seems really unlikely that it could be related to my PR.

**Comment by cruhland** [comment]
Ninja-merging this PR, as the tests did pass once (see status for commit `ab58211`).
